### PR TITLE
[Unittests] Remove Site and Region FKs from model classes and enforce Location FK not null.

### PR DESCRIPTION
**Pull request by HanlinMiao**
# Closes: #2957 
# What's Changed

Unittests for all apps after removing Site and Region FKs from model classes and enforcing Location FK not null.
- [x] Circuits
- [x] Core
- [x] DCIM
- [x] Extras
- [x] IPAM
- [x] Virtualization

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/core/tests/integration/test_filters.py [hunk 1]
```diff
@@ -13,8 +13,7 @@ class ListViewFilterTestCase(SeleniumTestCase):
     def setUp(self):
         super().setUp()
         self.login(self.user.username, self.password)
-        factory.RegionFactory.create_batch(15, has_parent=False)
-        factory.SiteFactory.create_batch(15, has_tenant=False)
+        factory.LocationFactory.create_batch(15, has_parent=True)
```
Comment (glennmatthews) [inline nautobot/core/tests/integration/test_filters.py:16]: Do we need `has_tenant=False` here?

File: nautobot/core/tests/integration/test_navbar.py [hunk 2]
```diff
@@ -10,9 +10,9 @@ class NavBarTestCase(SeleniumTestCase):
...
         "Organization": {
-            "Sites": {
-                "Sites": {
-                    "permission": "dcim.view_site",
+            "Locations": {
+                "Locations": {
+                    "permission": "dcim.view_location",
                     "buttons": ["Add", "Import"],
                 },
                 "Regions": {
```
Comment (glennmatthews) [inline nautobot/core/tests/integration/test_navbar.py:18]: remove?

File: nautobot/core/tests/test_filters.py [hunk 3]
```diff
@@ -741,21 +781,15 @@ def setUpTestData(cls):
...
-
-        cls.sites = (
-            dcim_models.Site.objects.filter(region=cls.regions[0]).first(),
-            dcim_models.Site.objects.filter(region=cls.regions[1]).first(),
-            dcim_models.Site.objects.filter(region=cls.regions[2]).first(),
-        )
-        cls.sites[0].asn = 65001
-        cls.sites[1].asn = 65101
-        cls.sites[2].asn = 65201
+        cls.locations = dcim_models.Location.objects.filter(parent__isnull=False)[:3]
```
Comment (glennmatthews) [inline nautobot/core/tests/test_filters.py:784]: Probably should also restrict by location-type or otherwise ensure that the selected locations are ones that actually permit Racks? (Since we're using `bulk_create` for the Racks below it's bypassing any validation that we have there.)

File: nautobot/dcim/forms.py [hunk 4]
```diff
@@ -4202,7 +4192,7 @@ class PowerFeedForm(NautobotModelForm):
     location = DynamicModelChoiceField(
         queryset=Location.objects.all(),
         required=False,
-        initial_params={"powerpanel": "$power_panel"},
+        initial_params={"powerpanels": "$power_panel"},
```
Comment (glennmatthews) [inline nautobot/dcim/forms.py:4195]: Hm. Not `power_panels`?

File: nautobot/dcim/signals.py [hunk 5]
```diff
@@ -66,53 +67,68 @@ def handle_rackgroup_location_change(instance, created, **kwargs):
...
+                    else:
+                        raise ValidationError(
+                            {
+                                f"location {instance.location.name}": "RackGroups may not associate to locations of type "
+                                f'"{instance.location.location_type}"'
+                            }
+                        )
+
+                if changed:
+                    rackgroup.save()
```
Comment (glennmatthews) [inline nautobot/dcim/signals.py:96]: Could this be simplified as follows here and below?

```suggestion
                if rackgroup.location not in descendants:
                    if not rack_groups_permitted:
                        raise ValidationError(
                            {
                                f"location {instance.location.name}": "RackGroups may not associate to locations of type "
                                f'"{instance.location.location_type}"'
                            }
                        )
                    rackgroup.location = instance.location
                    rackgroup.save()
```

File: nautobot/dcim/tests/test_models.py [hunk 6]
```diff
@@ -768,13 +622,19 @@ def setUp(self):
...
-            name="Root", status=self.device_status, location_type=self.location_type_1, site=self.site
+            name="Root", status=self.device_status, location_type=self.location_type_1
         )
         self.location_2 = Location.objects.create(
             name="Leaf", status=self.device_status, location_type=self.location_type_2, parent=self.location_1
         )
+        self.location_3 = Location.objects.create(
+            name="Device Allowed Location",
+            status=self.device_status,
+            location_type=self.location_type_3,
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:636]: Do we need to add the Device content-type to location_type_3?

File: nautobot/dcim/tests/test_models.py [hunk 7]
```diff
@@ -1338,20 +1164,6 @@ def test_tagged_vlan_raise_error_if_mode_not_set_to_tagged(self):
             err.exception.message_dict["tagged_vlans"][0], "Mode must be set to tagged when specifying tagged_vlans"
         )
 
-    def test_error_raised_when_adding_tagged_vlan_with_different_site_from_interface_parent_site(self):
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:1341]: Should this become an equivalent test for Location mismatch instead of being removed?

File: nautobot/extras/tests/test_customfields.py [hunk 8]
```diff
@@ -475,17 +476,22 @@ def setUpTestData(cls):
...
 
-        statuses = Status.objects.get_for_model(Site)
+        statuses = Status.objects.get_for_model(Location)
 
-        # Create some sites
-        cls.sites = (
-            Site.objects.create(name="Site 1", slug="site-1", status=statuses.get(slug="active")),
-            Site.objects.create(name="Site 2", slug="site-2", status=statuses.get(slug="active")),
+        # Create some locations
+        cls.lt = LocationType.objects.get(name="Campus")
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_customfields.py:482]: How badly will it break things if we update the location/location-type factories to create region-level and site-level locations? Obviously not part of this PR but something to think about...

**Review by glennmatthews — COMMENTED** [review]
Awesome work so far!

File: nautobot/core/tests/test_forms.py [hunk 9]
```diff
@@ -716,25 +716,27 @@ def test_dynamic_filter_form(self):
                     ("id", "Id"),
                     ("last_updated", "Last updated"),
                     ("latitude", "Latitude"),
-                    ("locations", "Locations within this Site (slugs or IDs)"),
+                    ("location_type", "Location type (slug or ID)"),
+                    ("content_type", "Location type Permitted object types"),
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/core/tests/test_forms.py:720]: I swap the SiteForm out for LocationForm so I think thats why maybe?

File: nautobot/dcim/api/serializers.py [hunk 10]
```diff
@@ -126,6 +126,7 @@
     NestedRearPortSerializer,
     NestedRearPortTemplateSerializer,
     NestedRegionSerializer,
+    NestedSiteSerializer,
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/dcim/api/serializers.py:129]: I deleted it in my master branch but you need it for the ?brief=true queries. I deleted it by mistake so I added it back for unittests to pass. It is not missing in `next`

File: nautobot/core/tests/test_forms.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/core/tests/test_forms.py:720]: Oh wait I misread it you mean the content_type?

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/tests/test_forms.py:720]: That's what I was thinking, but on reflection it makes sense. This is so I can for example get only Locations that permit Racks, etc.

File: nautobot/dcim/forms.py (hunk 4 above)
Comment (HanlinMiao) [reply nautobot/dcim/forms.py:4195]: I think `initial_params` are referencing the actual parent model field in this case `powerpanels`. Unlike `query_params` which are using the filterset field. I got a FieldError when I changed it to `power_panels`.
```
# Set initial value based on prescribed child fields (if not already set)
        if not self.initial and self.initial_params:
            filter_kwargs = {}
            for kwarg, child_field in self.initial_params.items():
                value = form.initial.get(child_field.lstrip("$"))
                if value:
                    filter_kwargs[kwarg] = value
            if filter_kwargs:
                self.initial = self.queryset.filter(**filter_kwargs).first()
```

Comment (HanlinMiao) [reply nautobot/dcim/forms.py:4195]: We should change it back when the dcim related names and FK issue is closed tho https://github.com/nautobot/nautobot/issues/3066. Leaving a note here

Comment (glennmatthews) [reply nautobot/dcim/forms.py:4195]: Got it. Thanks!

**Review by jathanism — APPROVED** [review]
I would like to see the tests passing but everything else here looks pretty straightforward.

File: examples/example_plugin/example_plugin/signals.py [hunk 11]
```diff
@@ -36,4 +36,4 @@ def nautobot_database_ready_callback(sender, *, apps, **kwargs):
             "default": "Default value",
         },
     )
-    field.content_types.set([ContentType.objects.get_for_model(Site)])
+    field.content_types.set([ContentType.objects.get_for_model(Location)])
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply examples/example_plugin/example_plugin/signals.py:39]: Being handled in #3256.

File: nautobot/core/tests/test_filters.py [hunk 12]
```diff
@@ -13,275 +14,318 @@
 from nautobot.dcim import choices as dcim_choices
 from nautobot.dcim import filters as dcim_filters
 from nautobot.dcim import models as dcim_models
-from nautobot.dcim.models import Device
+from nautobot.dcim.models import Device, Location, LocationType
```
Comment (glennmatthews) [inline nautobot/core/tests/test_filters.py:17]: Probably should just be using `dcim_models` throughout this file instead?

File: nautobot/core/tests/test_utils.py [hunk 13]
```diff
@@ -440,20 +443,20 @@ def test_get_filterset_parameter_form_field(self):
...
+            device_fields = ["cluster", "device_type", "location"]
             for field_name in device_fields:
                 form_field = filtering.get_filterset_parameter_form_field(dcim_models.Device, field_name)
                 self.assertIsInstance(form_field, forms.DynamicModelMultipleChoiceField)
 
         with self.subTest("Test ChoiceField"):
-            site_fields = ["has_locations", "has_circuit_terminations", "has_devices"]
-            for field_name in site_fields:
-                form_field = filtering.get_filterset_parameter_form_field(dcim_models.Site, field_name)
+            location_fields = ["has_devices"]
```
Comment (glennmatthews) [inline nautobot/core/tests/test_utils.py:457]: We'll eventually want to add back `has_circuit_terminations` won't we?

File: nautobot/dcim/signals.py [hunk 14]
```diff
@@ -66,53 +67,52 @@ def handle_rackgroup_location_change(instance, created, **kwargs):
...
     Note that this is non-trivial for Location changes, since a LocationType that can contain RackGroups
     may or may not be permitted to contain Racks or PowerPanels. If it's not permitted, rather than trying to search
-    through child locations to find the "right" one, the best we can do is simply to null out the location.
+    through child locations to find the "right" one, the best we can do is to raise to raise a ValidationError
+    and roll back the changes we made.
     """
-    if not created:
-        if instance.location is not None:
+    with transaction.atomic():
+        if not created:
```
Comment (glennmatthews) [inline nautobot/dcim/signals.py:74]: Would it make sense to instead do `if created: return` first to reduce the amount of nesting of code here and in the next function?

File: nautobot/dcim/tests/test_api.py [hunk 15]
```diff
@@ -362,16 +362,15 @@ def setUpTestData(cls):
...
+        cls.loc2 = Location.objects.create(name="RTP4E", location_type=cls.lt2, status=status_active, parent=cls.loc1)
+        cls.loc3 = Location.objects.create(name="RTP4E-3", location_type=cls.lt3, status=status_active, parent=cls.loc2)
+        cls.loc4 = Location.objects.create(
+            name="RTP4E-3-0101", location_type=cls.lt4, status=status_active, parent=cls.loc3, tenant=tenant
         )
-        for loc in [loc1, loc2, loc3, loc4]:
+        for loc in [cls.loc1, cls.loc2, cls.loc3, cls.loc4]:
             loc.validated_save()
 
         # FIXME(jathan): The writable serializer for `Device.status` takes the
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_api.py:376]: Gonna have some fun merge conflicts to resolve now that #3225 is in `next`... sorry!

File: nautobot/dcim/tests/test_models.py [hunk 16]
```diff
@@ -319,13 +266,9 @@ def test_change_rackgroup_location_children_not_permitted(self):
...
-        self.rackgroup_a1.save()
-
-        self.assertEqual(RackGroup.objects.get(pk=self.rackgroup_a1.pk).location, location_c)
-        self.assertEqual(RackGroup.objects.get(pk=self.rackgroup_a2.pk).location, location_c)
-        self.assertEqual(Rack.objects.get(pk=self.rack1.pk).location, None)
-        self.assertEqual(Rack.objects.get(pk=self.rack2.pk).location, None)
-        self.assertEqual(PowerPanel.objects.get(pk=self.powerpanel1.pk).location, None)
+        with self.assertRaises(ValidationError) as cm:
+            self.rackgroup_a1.save()
+        self.assertIn(f'Racks may not associate to locations of type "{location_type_c}"', str(cm.exception))
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:271]: Good to assert that none of the RackGroup/Rack/PowerPanel objects got updated to the new location?

File: nautobot/dcim/tests/test_models.py [hunk 17]
```diff
@@ -548,28 +440,16 @@ def test_change_rack_location_devices_not_permitted(self):
...
-            rack.validated_save()
-        self.assertIn("Assigned rack group must belong to parent site", str(cm.exception))
-
-        # Rack location and site must match
-        rack = Rack(name="Rack", site=self.site2, location=self.location1, status=self.status)
         with self.assertRaises(ValidationError) as cm:
-            rack.validated_save()
-        self.assertIn(f'Location "Location1" does not belong to site "{self.site2.name}"', str(cm.exception))
+            self.rack.save()
+        self.assertIn(f'Devices may not associate to locations of type "{location_type_b}"', str(cm.exception))
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:447]: Similarly, anything we should assert was not changed as a result?

File: nautobot/extras/management/__init__.py [hunk 18]
```diff
@@ -73,10 +74,10 @@
     "planned": "Unit has been planned",
     "provisioning": "Circuit is being provisioned",
     "reserved": "Unit is reserved",
-    "retired": "Location has been retired",
+    "retired": "Site or Location has been retired",
     "slaac": "Dynamically assigned IPv6 address",
     "staged": "Unit has been staged",
-    "staging": "Location is in the process of being staged",
+    "staging": "Site or Location is in the process of being staged",
```
Comment (glennmatthews) [inline nautobot/extras/management/__init__.py:80]: Maybe I asked this already - but why the additions in this file?

File: nautobot/extras/tests/test_views.py [hunk 19]
```diff
@@ -14,7 +14,7 @@
 from nautobot.core.models.fields import slugify_dashes_to_underscores
 from nautobot.core.testing import ViewTestCases, TestCase, extract_page_body, extract_form_failures
 from nautobot.core.testing.utils import disable_warnings, post_data
-from nautobot.dcim.models import ConsolePort, Device, DeviceType, Interface, Manufacturer, Site
+from nautobot.dcim.models import ConsolePort, Device, DeviceType, Interface, Manufacturer, Location, LocationType
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_views.py:17]: nit: ordering

```suggestion
from nautobot.dcim.models import ConsolePort, Device, DeviceType, Interface, Location, LocationType, Manufacturer
```

**Review by glennmatthews — APPROVED** [review]
